File: cmd/exs-node/oracle.go
package main

import (
	"fmt"
	"strings"

	"github.com/spf13/cobra"
)

var oracleCmd = &cobra.Command{
	Use:   "oracle",
	Short: "Oracle operations (Protocol intelligence)",
	Long: `Consult the Excalibur Oracle for protocol guidance and intelligence.
	
Features from Knights' Round Table Oracle:
  • Protocol guidance and mining help
  • Divination for forge outcomes
  • Treasury and revenue insights
  • Technical specifications`,
}

var oracleAskCmd = &cobra.Command{
	Use:   "ask [question...]",
	Short: "Ask the oracle a question",
	Args:  cobra.MinimumNArgs(1),
	Run: func(cmd *cobra.Command, args []string) {
		question := strings.Join(args, " ")
		
		fmt.Println("🔮 Consulting the Oracle...")
		fmt.Println("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━")
		fmt.Printf("Question: %s\n", question)
		fmt.Println()
		
		response := consultOracle(question)
		
		fmt.Println("Oracle's Wisdom:")
		fmt.Println(response.Wisdom)
		
		if len(response.Details) > 0 {
			fmt.Println("\nDetails:")
			for key, value := range response.Details {
				fmt.Printf("  %s: %s\n", key, value)
			}
		}
	},
}

var oracleDivinationCmd = &cobra.Command{
	Use:   "divine",
	Short: "Get divination for forge outcome",
	Run: func(cmd *cobra.Command, args []string) {
		fmt.Println("🔮 Performing Divination...")
		fmt.Println("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━")
		
		// Simulate divination
		divination := performDivination()
		
		fmt.Println(divination.Message)
		fmt.Printf("\nSuccess Probability: %s\n", divination.Probability)
		fmt.Printf("Recommended Action: %s\n", divination.Recommendation)
		
		if len(divination.Warnings) > 0 {
			fmt.Println("\n⚠️  Warnings:")
			for _, warning := range divination.Warnings {
				fmt.Printf("  • %s\n", warning)
			}
		}
	},
}

var oracleQuickCmd = &cobra.Command{
	Use:   "quick [topic]",
	Short: "Quick oracle queries (mining, forge, treasury, rewards)",
	Args:  cobra.ExactArgs(1),
	Run: func(cmd *cobra.Command, args []string) {
		topic := strings.ToLower(args[0])
		
		fmt.Println("🔮 Quick Oracle Response")
		fmt.Println("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━")
		
		switch topic {
		case "mining":
			fmt.Println("Mining Guidance:")
			fmt.Println("  Use Ω′ Δ18 Tetra-PoW with 128 rounds")
			fmt.Println("  Algorithm: 128-round unrolled nonlinear hash")
			fmt.Println("  Difficulty: 4 leading zero bytes")
			fmt.Println("  Command: exs-node mine start --address <addr>")
		case "forge":
			fmt.Println("Forge Process:")
			fmt.Println("  Every successful forge rewards 50 $EXS")
			fmt.Println("  15% (7.5 $EXS) goes to treasury")
			fmt.Println("  Use: exs-node forge start --address <addr>")
		case "treasury":
			fmt.Println("Treasury Information:")
			fmt.Println("  Rolling 12-month release with CLTV time-locks")
			fmt.Println("  9 revenue streams generating yield")
			fmt.Println("  Use: exs-node revenue show")
		case "rewards":
			fmt.Println("Reward System:")
			fmt.Println("  Forge reward: 50 EXS")
			fmt.Println("  Smooth exponential halving over 1000 forges")
			fmt.Println("  Tail emission: 0.1 EXS minimum forever")
		default:
			fmt.Printf("Unknown topic: %s\n", topic)
			fmt.Println("Available topics: mining, forge, treasury, rewards")
		}
	},
}

type OracleResponse struct {
	Wisdom  string
	Details map[string]string
}

type Divination struct {
	Message        string
	Probability    string
	Recommendation string
	Warnings       []string
}

func consultOracle(question string) OracleResponse {
	questionLower := strings.ToLower(question)
	
	if strings.Contains(questionLower, "mine") || strings.Contains(questionLower, "mining") {
		return OracleResponse{
			Wisdom: "As Arthur proved his worth by drawing Excalibur, so must miners prove theirs through the Ω′ Δ18 forge.",
			Details: map[string]string{
				"Algorithm":     "Ω′ Δ18 (128-round unrolled nonlinear hash)",
				"Difficulty":    "4 leading zero bytes",
				"Quantum-Hard":  "HPP-1 with 600,000 PBKDF2 rounds",
				"Requirements":  "13-word axiom, Valid nonce, Computational power",
			},
		}
	} else if strings.Contains(questionLower, "forge") || strings.Contains(questionLower, "forging") {
		return OracleResponse{
			Wisdom: "Every successful forge echoes through Camelot, rewarding the worthy with 50 $EXS.",
			Details: map[string]string{
				"Forge Reward": "50 $EXS per successful forge",
				"Treasury":     "15% (7.5 $EXS) to treasury",
				"Process":      "Verify axiom → Draw sword → Mine 128 rounds → Receive P2TR vault",
			},
		}
	} else if strings.Contains(questionLower, "treasury") {
		return OracleResponse{
			Wisdom: "The treasury sustains itself through nine mystical streams, each flowing with perpetual abundance.",
			Details: map[string]string{
				"Revenue Streams": "9 sources (mining, staking, DeFi, MEV, etc.)",
				"Release":         "12-month rolling with CLTV time-locks",
				"Allocation":      "15% of all forge rewards",
			},
		}
	} else if strings.Contains(questionLower, "taproot") || strings.Contains(questionLower, "p2tr") {
		return OracleResponse{
			Wisdom: "Taproot vaults conceal their secrets behind quantum-hardened prophecies, unlinkable and eternal.",
			Details: map[string]string{
				"Address Type": "P2TR (Pay-to-Taproot)",
				"Encoding":     "Bech32m",
				"Privacy":      "Unlinkable outputs via 13-word axiom",
				"Quantum-Hard": "HPP-1 key derivation",
			},
		}
	}
	
	return OracleResponse{
		Wisdom: "The Oracle sees many paths. Ask about mining, forge, treasury, or taproot for specific guidance.",
		Details: map[string]string{
			"Available Topics": "mining, forge, treasury, taproot, rewards, protocol",
		},
	}
}

func performDivination() Divination {
	return Divination{
		Message:        "The stars align favorably for your forge. The Ω′ Δ18 algorithm flows through your hardware like Excalibur through stone.",
		Probability:    "High (based on current difficulty and hashrate)",
		Recommendation: "Proceed with forge using optimal thread count",
		Warnings: []string{
			"Ensure adequate cooling for extended mining",
			"Verify axiom before starting forge",
			"Monitor power consumption",
		},
	}
}

func init() {
	oracleCmd.AddCommand(
		oracleAskCmd,
		oracleDivinationCmd,
		oracleQuickCmd,
	)
	
	rootCmd.AddCommand(oracleCmd)
}

File: web/knights-round-table/oracle.js
// Excalibur Oracle - Web Portal Interface
// Connects to the Oracle backend for protocol intelligence

// Protocol constants
const PROTOCOL_AXIOM = "sword legend pull magic kingdom artist stone destroy forget fire steel honey question";
const TAPROOT_ADDRESS = "bc1pql83shz0m4znewzk82u2k5mdgeh94r3c8ks9ws00m4dm26qnjvyq0prk4n";

let queriesProcessed = 0;

// Initialize oracle interface
document.addEventListener('DOMContentLoaded', function() {
    // Display Taproot address
    document.getElementById('taproot-addr').textContent = TAPROOT_ADDRESS.substring(0, 20) + '...';
    
    // Set up event listeners
    document.getElementById('ask-oracle-btn').addEventListener('click', askOracle);
    document.getElementById('divination-btn').addEventListener('click', getDivination);
    
    // Allow Enter key in textarea (Shift+Enter for new line)
    document.getElementById('oracle-query').addEventListener('keypress', function(e) {
        if (e.key === 'Enter' && !e.shiftKey) {
            e.preventDefault();
            askOracle();
        }
    });
});

// Sanitize user input
function sanitizeInput(input) {
    const div = document.createElement('div');
    div.textContent = input;
    return div.innerHTML;
}

// Quick query function
function quickQuery(query) {
    document.getElementById('oracle-query').value = query;
    askOracle();
}

// Ask the oracle
async function askOracle() {
    const queryInput = document.getElementById('oracle-query');
    const rawQuery = queryInput.value.trim();
    
    if (!rawQuery) {
        alert('Please enter a question for the oracle.');
        return;
    }
    
    if (rawQuery.length > 500) {
        alert('Question too long. Please keep it under 500 characters.');
        return;
    }
    
    const query = sanitizeInput(rawQuery);
    const button = document.getElementById('ask-oracle-btn');
    const responseDiv = document.getElementById('oracle-response');
    
    // Disable button during processing
    button.disabled = true;
    button.textContent = '🔮 Consulting Oracle...';
    
    // Show loading
    showLoading(responseDiv);
    
    // Simulate oracle consultation (in production, this would call the backend)
    setTimeout(() => {
        const response = consultOracleLocal(query);
        displayOracleResponse(response);
        
        button.disabled = false;
        button.textContent = '🔮 Consult the Oracle';
        
        queriesProcessed++;
        document.getElementById('queries-count').textContent = queriesProcessed;
    }, 1500);
}

// Local oracle simulation (mimics the Python oracle)
function consultOracleLocal(query) {
    const queryLower = query.toLowerCase();
    
    // Determine query category and provide response
    if (queryLower.includes('mine') || queryLower.includes('mining')) {
        return {
            category: 'Mining Guidance',
            wisdom: 'As Arthur proved his worth by drawing Excalibur, so must miners prove theirs through the Ω′ Δ18 forge.',
            details: {
                overview: 'Use the Ω′ Δ18 Tetra-PoW miner with 128 rounds',
                algorithm: 'Ω′ Δ18 (128-round unrolled nonlinear hash)',
                difficulty: '4 leading zero bytes',
                command: 'python3 pkg/miner/tetra_pow_miner.py --axiom "[13 words]" --difficulty 4',
                requirements: ['13-word axiom', 'Valid nonce', 'Computational power']
            }
        };
    } else if (queryLower.includes('forge') || queryLower.includes('forging')) {
        return {
            category: 'Forge Process',
            wisdom: 'Every successful forge echoes through Camelot, rewarding the worthy with 50 $EXS.',
            details: {
                overview: 'Forge $EXS tokens by mining valid proofs',
                reward: '50 $EXS per successful forge',
                miner_receives: '49.5 $EXS (after 1% treasury fee)',
                fees: {
                    treasury: '0.5 $EXS (1%)',
                    btc_forge_fee: '0.0001 BTC'
                },
                process: [
                    '1. Enter the 13-word axiom',
                    '2. Mine for a valid nonce',
                    '3. Submit proof',
                    '4. Receive reward'
                ]
            }
        };
    } else if (queryLower.includes('vault') || queryLower.includes('address')) {
        return {
            category: 'Vault Information',
            wisdom: 'Each vault is a Taproot mystery, deterministic yet unlinkable, like the Lady\'s lake.',
            details: {
                overview: 'Taproot vaults are generated deterministically',
                type: 'P2TR (Pay-to-Taproot)',
                standard: 'BIP-86',
                generation: 'axiom + nonce → HPP-1 key → Taproot address',
                security: '600,000 PBKDF2 iterations',
                tweak: 'Custom 13-word axiom as Taproot tweak'
            }
        };
    } else if (queryLower.includes('treasury') || queryLower.includes('admin')) {
        return {
            category: 'Treasury Control',
            wisdom: 'Merlin guards the treasury with 1.2 million magical rounds, twice the strength of ordinary keys.',
            details: {
                overview: 'Treasury admin credentials use enhanced security',
                security: '1.2 million PBKDF2 iterations (2x forge keys)',
                access: 'Merlin\'s Portal with MERLIN-{id} credentials',
                command: 'python3 forge_treasury_key.py',
                control: 'Full $EXS Treasury management',
                fee_collection: '1% of all forge rewards'
            }
        };
    } else if (queryLower.includes('axiom') || queryLower.includes('prophecy')) {
        return {
            category: 'Protocol Axiom',
            wisdom: 'The 13 words of power bind the protocol, each word a knight at the Round Table.',
            details: {
                axiom: PROTOCOL_AXIOM,
                words: 13,
                hash: computeAxiomHash(PROTOCOL_AXIOM),
                importance: 'Foundation of all vault generation and proofs',
                usage: 'Required for all mining and forge operations',
                immutable: 'Canonical and unchangeable'
            }
        };
    } else if (queryLower.includes('difficulty') || queryLower.includes('target')) {
        return {
            category: 'Difficulty Requirements',
            wisdom: 'Four leading zeros mark the challenge, a dragon\'s gate that few may pass.',
            details: {
                difficulty: '4 leading zero bytes',
                hex_representation: '00000000...',
                bits: '32 zero bits',
                algorithm: 'Ω′ Δ18 Tetra-PoW',
                rounds: 128,
                adjustment: 'Manual by protocol architect'
            }
        };
    } else if (queryLower.includes('taproot') || queryLower.includes('address')) {
        return {
            category: 'Taproot Integration',
            wisdom: 'The sword remains in the stone until cryptographic proof is shown.',
            details: {
                real_address: TAPROOT_ADDRESS,
                verification: 'CRYPTOGRAPHIC_REALITY_VERIFIED',
                derivation: 'BIP32/BIP86',
                type: 'P2TR (witness version 1)',
                tweak_method: 'SHA256(internalKey || prophecyHash)'
            }
        };
    } else if (queryLower.includes('satoshi') || queryLower.includes('nakamoto') || queryLower.includes('bitcoin philosophy') || queryLower.includes('genesis') || queryLower.includes('inscription')) {
        return {
            category: 'Satoshi Nakamoto\'s Wisdom',
            wisdom: 'The root problem with conventional currency is all the trust that\'s required to make it work. We eliminate trust through cryptographic proof.',
            details: {
                philosophy: 'Satoshi Nakamoto created Bitcoin to enable peer-to-peer electronic cash without trusted third parties - like sovereign crews sailing beyond admiralty jurisdiction',
                key_principles: [
                    'Decentralization: No single point of failure or control - captains elected and deposed by vote',
                    'Trustlessness: Don\'t trust, verify through mathematics - Articles signed in code, not enforced by kings',
                    'Proof-of-Work: Computational consensus solves Byzantine Generals\' Problem - like Articles binding independent crews',
                    'Fixed Supply: 21 million coin cap creates sound money - £100,000 fortunes built through controlled supply',
                    'Censorship Resistance: No entity can block transactions - no empire could stop sovereign ships flying independent colors',
                    'Peer-to-Peer: Direct value transfer without intermediaries - prize captures bypassed admiralty courts, no king\'s cut'
                ],
                satoshi_quotes: 'It is better to have a system where no trust is needed at all',
                genesis_proof: 'Genesis block 000000000019d6689c085ae165831e934ff763ae46a2a6c172b3f1b60a8ce26f marks provable time. The Times 03/Jan/2009 Chancellor on brink of second bailout - Satoshi\'s inscription of defiance.',
                provable_fairness: 'Genesis roll 0.28 verified via HMAC-SHA512. Rolls 42.04, 23.01, 31.55, 54.36, 20.22 - all verifiable against block hash. Mathematics enforces fairness.',
                inscription_layer: 'Block 840,001+ enables ordinal inscriptions. Transaction f4293583b with 3,500+ confirmations - permanent cultural layer on Bitcoin. Even 600 sats from faucets can fund sovereignty.',
                hidden_axiom: 'Democratic crews governed by published Articles. From Caribbean to South China Sea - 80,000 strong in confederation. Reputation as an asset, psychological warfare preserved cargo value. Smart operators reinvested in legitimate empires, retired as wealthy governors.',
                dust_to_power: 'Dust collection demonstrates permissionless nature - anyone can participate. 600 sats aggregated funds inscriptions. No minimum balance, no gatekeepers.',
                relevance_to_exs: 'Excalibur $EXS inherits these principles with quantum-hardened security. Self-custody means being your own bank - or your own sovereign ship on the high seas.'
            }
        };
    } else if (queryLower.includes('decentraliz') || queryLower.includes('trustless') || queryLower.includes('freedom') || queryLower.includes('provable')) {
        return {
            category: 'Decentralization & Trustlessness',
            wisdom: 'Camelot has no king but the protocol itself. Satoshi showed us a system where no trust is needed at all.',
            details: {
                overview: 'Decentralization removes single points of failure - like independent ships as sovereign nations',
                no_central_authority: 'No single entity controls the network - captains elected and deposed by crew vote, no crown above the crew',
                permissionless: 'Anyone can participate without permission - any sailor could join the account, meritocracy over birthright',
                trustless_verification: 'Mathematical proofs replace human trust - Articles signed in blood, enforced by consensus',
                censorship_resistant: 'No one can block valid transactions - no empire could stop ships flying sovereign colors',
                self_custody: 'Be your own bank, control your own keys - buried treasure where no sovereign could seize it',
                peer_to_peer: 'Direct value transfer without intermediaries - prize captures bypassed king\'s taxes',
                satoshi_principle: 'Peer-to-peer electronic cash allows direct payments without financial institutions',
                hidden_axiom: '1,800 ships strong in confederation. Controlled entire trade routes through coordination. Economies of scale beyond empires. Fair shares for all crew - from captain (1.5 shares) to powder monkey (0.25 shares).',
                sovereignty: 'Each ship an independent nation on the high seas - insurance, democracy, freedom from chains'
            }
        };
    } else if (queryLower.includes('sound money') || queryLower.includes('scarcity') || queryLower.includes('wealth')) {
        return {
            category: 'Sound Money & Scarcity',
            wisdom: '21 million $EXS total, like Bitcoin\'s 21 million coins. Absolute scarcity creates sound money.',
            details: {
                total_supply: '21,000,000 $EXS (absolute cap)',
                forge_reward: '50 $EXS per forge',
                scarcity_principle: 'Fixed supply prevents inflation and debasement - through systematic targeting of richest routes',
                store_of_value: 'Digital gold for preserving wealth over time - buried treasure preserved across centuries, beyond sovereign reach',
                satoshi_insight: 'Steady addition of coins is analogous to gold miners adding gold to circulation - or prize captures from monopolistic routes',
                hidden_axiom: 'Systematic approach: captured over 400 ships worth millions. Protection economies and taxation systems. $200M+ raids, vanished with fortune, never caught. Smuggling operations avoiding customs duties. Reputation more valuable than gold itself.',
                sound_money_properties: ['Scarcity', 'Durability', 'Divisibility', 'Portability', 'Fungibility'],
                wealth_creation: 'Value creation outside the system - smart operators reinvested in legitimate businesses, retired wealthy governors'
            }
        };
    } else {
        return {
            category: 'General Protocol Wisdom',
            wisdom: 'In the land of Excalibur, cryptographic truth reigns supreme.',
            details: {
                protocol: 'Excalibur $EXS',
                total_supply: '21,000,000 $EXS',
                forge_reward: '50 $EXS',
                distribution: {
                    miners: '60%',
                    treasury: '15%',
                    liquidity: '20%',
                    airdrop: '5%'
                },
                status: 'OPERATIONAL',
                ask_specific: 'Try asking about: mining, forge, vault, treasury, axiom, difficulty, Satoshi Nakamoto, decentralization, sound money, freedom, wealth creation'
            }
        };
    }
}

// Display oracle response
function displayOracleResponse(response) {
    const responseDiv = document.getElementById('oracle-response');
    const wisdomDiv = document.getElementById('oracle-wisdom');
    const categoryDiv = document.getElementById('oracle-category');
    const detailsDiv = document.getElementById('oracle-details');
    
    // Set wisdom
    wisdomDiv.textContent = '💭 ' + response.wisdom;
    
    // Set category
    categoryDiv.textContent = '📂 Category: ' + response.category;
    
    // Format details
    let detailsHTML = '<div style="margin-top: 15px;">';
    
    if (typeof response.details === 'object') {
        for (const [key, value] of Object.entries(response.details)) {
            if (typeof value === 'object' && !Array.isArray(value)) {
                detailsHTML += `<p style="margin-top: 10px;"><strong>${formatKey(key)}:</strong></p>`;
                detailsHTML += '<ul style="margin-left: 20px;">';
                for (const [subkey, subvalue] of Object.entries(value)) {
                    detailsHTML += `<li>${formatKey(subkey)}: ${subvalue}</li>`;
                }
                detailsHTML += '</ul>';
            } else if (Array.isArray(value)) {
                detailsHTML += `<p style="margin-top: 10px;"><strong>${formatKey(key)}:</strong></p>`;
                detailsHTML += '<ul style="margin-left: 20px;">';
                value.forEach(item => {
                    detailsHTML += `<li>${item}</li>`;
                });
                detailsHTML += '</ul>';
            } else {
                detailsHTML += `<p style="margin-top: 5px;"><strong>${formatKey(key)}:</strong> ${value}</p>`;
            }
        }
    }
    
    detailsHTML += '</div>';
    detailsDiv.innerHTML = detailsHTML;
    
    // Show response
    responseDiv.classList.add('active');
}

// Get oracle divination
async function getDivination() {
    const button = document.getElementById('divination-btn');
    const resultDiv = document.getElementById('divination-result');
    const textDiv = document.getElementById('divination-text');
    const statusDiv = document.getElementById('divination-status');
    
    button.disabled = true;
    button.textContent = '✨ Channeling Oracle...';
    
    // Simulate divination
    setTimeout(() => {
        const divinations = [
            "The Round Table awaits worthy knights to forge their destiny.",
            "Camelot's treasury grows with each successful proof of work.",
            "The Lady of the Lake whispers secrets through Taproot addresses.",
            "Merlin's magic flows through 1.2 million iterations of power.",
            "The sword remains in the stone until cryptographic proof is shown.",
            "Four zeros mark the dragon's challenge - prove your worth.",
            "The 13 words bind the protocol in eternal cryptographic truth.",
            "Each forge strengthens the kingdom, each miner a knight of honor.",
            "The oracle sees all transactions, immutable and eternal.",
            "Wisdom flows from the blockchain like water from Avalon's springs.",
            "Satoshi taught us: no trust is needed when mathematics enforces consensus.",
            "Like sovereign ships beyond admiralty jurisdiction, crypto sails beyond central authority.",
            "Democratic crews voted on decisions - captains elected by consensus, not appointed by crowns.",
            "Reputation as an asset - psychological warfare preserved value better than violence.",
            "From Caribbean to South China Sea - coordination at scale creates power beyond empires.",
            "Genesis block 000000000019d6 marks provable time - The Times inscription eternal.",
            "Even 600 sats collected can fund inscriptions - permissionless sovereignty for all."
        ];
        
        const index = Math.floor(Math.random() * divinations.length);
        const divination = divinations[index];
        
        textDiv.textContent = '"' + divination + '"';
        statusDiv.textContent = '🔮 Oracle Status: OPERATIONAL | Protocol: ACTIVE';
        
        resultDiv.style.display = 'block';
        
        button.disabled = false;
        button.textContent = '✨ Receive Divination';
        
        queriesProcessed++;
        document.getElementById('queries-count').textContent = queriesProcessed;
    }, 1000);
}

// Show loading state
function showLoading(container) {
    container.innerHTML = '<div class="loading"><div class="spinner"></div><p>Consulting the oracle...</p></div>';
    container.classList.add('active');
}

// Format key for display
function formatKey(key) {
    return key.replace(/_/g, ' ').replace(/\b\w/g, l => l.toUpperCase());
}

// Compute axiom hash (SHA-256)
function computeAxiomHash(axiom) {
    // Simple hash visualization (in production, use proper SHA-256)
    let hash = 0;
    for (let i = 0; i < axiom.length; i++) {
        const char = axiom.charCodeAt(i);
        hash = ((hash << 5) - hash) + char;
        hash = hash & hash;
    }
    return Math.abs(hash).toString(16).padStart(64, '0').substring(0, 64);
}
